src/Grpc.Net.Client/Balancer/Internal/BalancerAddressEqualityComparer.cs

```diff
@@ -44,6 +44,46 @@ public bool Equals(BalancerAddress? x, BalancerAddress? y)
             return false;
         }
 
+        var xAttributes = x._attributes?._attributes;
+        var yAttributes = y._attributes?._attributes;
+        if (!AttributesEqual(xAttributes, yAttributes))
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private bool AttributesEqual(Dictionary<string, object?>? x, Dictionary<string, object?>? y)
```

Comment:
nit: maybe a `ReferenceEquals` check first?

nit: as a general guidance, I've been trying to push people to always prefer `is null` / `is not null` when performing `null` tests - it isn't a big thing here, but it can make a big difference when custom equality operators are in play (for example `string`); entirely up to you, but I see only upside to `is null` vs `== null` etc

Reply:
> nit: maybe a ReferenceEquals check first (replaces both-null)?

Done.

> is null

The types here are all known and don't have custom equals. I agree in principle, but not enough to change.